### DISCUSSION TITLE
API documentation & automation

### DISCUSSION THREAD

**Discussion by vinukumar-vs** [Sunbird-Knowlg/Community#35] [General Discussions]
To increase the stability & reliability of the microservice, the API tests automation & documentation plays a very important role. So as part of buildingblocks, we are planning to upgrade the microservice API documentation & automation of API tests.

Please find the below document for more details that each Sunbird buildingblock should adopt.

Document: [API documentation & automation](https://docs.google.com/document/d/11QDIMddIJn_xwvm65xA6-QIGnq0yBnl-GPvsZzMyBec/edit?usp=sharing)

Please share your valuable feedback & suggestions by commenting on the above document.
@pramodkvarma,  @vrayulu, @SanthoshVasabhaktula, @rhwarrier,
@rahul-kankaria, @ChakshuGautam, @kiranharidas187, @pritamps, @chitranshu-keshav @rahul101001000 @coolbung
@surendrasinghs, @pallakartheekreddy, @krgauraw, @luckynvdu @reshmi-nair, @sharathkashyap, @gandham-santhosh, @kameshbhr, @AmiableAnil, @Jayaprakash8887, @anandp504,
@surendrasinghs

Shikshalokam: @vijiurs @aishwaryashikshalokam @Prateek-slokam @aks30 @varsha-sl @MeghnaSL
Samagra: @Prerna789 @aayushkedia95
SunbirdCB: @pkranga
Avanti: @pritamps @karn-nahata
DC's: @SanthoshVasabhaktula @vrayulu @maheshkumargangula
Sunbird QA: @ashokreddy1208
@gandham-santhosh
@kameshbhr @mmanimegalai
@madhucr @alok-os

**Comment by rhwarrier** [comment]
Thanks for the doc, @Vinukumar-vs

Please note that the doc lists out the key outputs that are expected for all contributions to the building block - this will have to be adhered to as a guideline when making contributions/ building for the Knowlg block. The team has also tried out different tools to achieve the requisite outcomes, and the doc also lists out the benefits of some of these. 

Requesting community members to weigh in on the suggestions so that we can formulate a generalised set of guidelines for all building blocks.

**Comment by vinukumar-vs** [comment]
For Sunbird-Knowlg building block we are going to use Postman tool for API documentation & Automation tests. We started using this for 
DIAL service(work inprogress). Please find the below for reference

[DIAL service API documentation](https://documenter.getpostman.com/view/4214340/UVsMv6Kd)
[DIAL service automation tests](https://www.postman.com/knowlg-sunbird/workspace/dial/documentation/4214340-c506c361-8f86-45c3-999d-9e6ed7a2e28d)

All Sunbird-knowlg microservices will get updated to the below Postman public API network
Postman Publish API Network:  [Sunbird-Knowlg](https://www.postman.com/knowlg-sunbird)

So going forward every PR of sunbird-knowlg mircorservices should have updated/new addedPostman collection(used for both API documentation & Automation). 

@pramodkvarma @SanthoshVasabhaktula @vrayulu @maheshkumargangula Please share your feedback & suggestions.

**Comment by rahul101001000** [comment]
Added comments to the doc

**Comment by vinukumar-vs** [comment]
We have noticed couple of changes while working on Automation tests

- Some of the content identifiers that we use for testing may not work for different environments. Because those content identifiers may not exist. 
  Solution: 
  1. Define all the required content identifiers, Collection Identifiers, Collection units etc in the environment variables. 
      These variables can be updated/configured based on the environments in which we want to execute the tests.
  2. Always create the new collection or content & store it in global level variables. Use these global variables where ever we need them in the test cases. 

- As a standard, each test case should be independently executable. 
  Solution:
  We are not following this approach at present. Our tests should be executed in the same order/sequence. Because we will be referring the response of previous tests instead of creating the required data again for the next test cases.(just to avoid the creation of required data for every tests)
  
  Please share if you find any better solutions or any suggestions
  @vrayulu @maheshkumargangula
